File: 306.additive-number.go
/*
 * @lc app=leetcode id=306 lang=golang
 *
 * [306] Additive Number
 */

// @lc code=start
import (
	"strconv"
)

func isAdditiveNumber(num string) bool {
	if len(num) < 3 {
		return false
	}
	// start from the first 2 numbers
	// special case: number starting with 0 is invalid in this problem
	for i := 0; i < len(num)-2; i++ {
		if num[0] == '0' && i > 0 {
			break
		}
		cur, _ := strconv.Atoi(num[:i+1])
		for j := i + 1; j < len(num)-1; j++ {
			if num[i+1] == '0' && j > i+1 {
				break
			}
			next, _ := strconv.Atoi(num[i+1 : j+1])
			if find(cur, next, num[j+1:len(num)]) {
				
				// fmt.Printf("%v %v", cur, next)
				return true
			}
		}

	}
	return false
}

func find(first int, second int, num string) bool {
	/*
		description
			find the third number which will make the whole string an additive number
		first: the first number
		second: the second number
		num: the given string
	*/
	if len(num) == 0 {
		return true
	}
	third := first + second
	third_str := strconv.Itoa(third)
	if strings.HasPrefix(num, third_str) {
		if find(second, third, num[len(third_str):len(num)]) {
			return true
		}
	}
	return false

}

// @lc code=end

File: 306.additive-number.java
/*
 * @lc app=leetcode id=306 lang=java
 *
 * [306] Additive Number
 */

// @lc code=start
class Solution {
    public boolean isAdditiveNumber(String num) {
        // get the first 2 numbers
        int N = num.length();
        for(int i=1; i<N-1; i++) {
            if(num.charAt(0)=='0'&&i>1) {
                break;
            }
            long firstNum = Long.valueOf(num.substring(0, i));
            for(int j=i+1; j<N; j++) {
                if(num.charAt(i)=='0'&&j>i+1) {
                    break;
                }
                long secondNum = Long.valueOf(num.substring(i, j));
                if(track(firstNum, secondNum, num, j)) {
                    return true;
                }
            }
        }
        return false;
    }

    // use the first numbers to check whether the rest of num string is additive number
    // int index from which the num string will be checked
    public boolean track(long firstNum, long secondNum, String num, int index) {
        long thirdNum = firstNum + secondNum;
        String thirdStr = String.valueOf(thirdNum);
        if(num.startsWith(thirdStr, index)) {
            if(thirdStr.length()+index==num.length()) {
                return true;
            } else {
                return track(secondNum, thirdNum, num, index+thirdStr.length());
            }
        } else {
            return false;
        }
    }
}
// @lc code=end

